crates/shared/src/caching_balance_fetcher.rs

```diff
@@ -0,0 +1,146 @@
+        });
+        CacheResponse {
+            cached,
+            missing,
+            block,
+        }
+    }
+
+    /// Updates the cache if the it doesn't contain more recent data.
+    fn update_balance_cache(
+        cache: &RwLock<BalanceCache>,
+        start_block: u64,
+        updates: Vec<(Query, U256)>,
+    ) {
+        let mut write_lock = cache.write().unwrap();
+        if write_lock.block > start_block {
+            // Newer data might already be availble which we don't want to overwrite.
+            return;
+        }
+        write_lock.data.extend(updates);
+    }
+
+    /// Spawns task that refreshes the cached balances on every new block.
+    pub fn spawn_background_task(&self, block_stream: CurrentBlockStream) {
+        let inner = self.inner.clone();
+        let cache = self.balance_cache.clone();
+        let mut stream = into_stream(block_stream);
+
+        let task = async move {
+            while let Some(block) = stream.next().await {
+                // invalidate cache immediately
+                let old_cache = {
+                    let empty_cache = BalanceCache {
+                        block: block.number,
+                        data: Default::default(),
+                    };
+                    let mut old_cache = cache.write().unwrap();
+                    std::mem::replace(&mut *old_cache, empty_cache)
+                };
+
+                let queries: Vec<_> = old_cache.data.into_keys().collect();
+                let results = inner.get_balances(&queries).await;
+                let updates = queries
+                    .into_iter()
+                    .zip(results.into_iter())
+                    .filter_map(|(query, result)| Some((query, result.ok()?)))
+                    .collect();
+                Self::update_balance_cache(&cache, block.number, updates);
+            }
+            tracing::error!("block stream terminated unexpectedly");
+        };
+        tokio::spawn(task.instrument(tracing::info_span!("balance_cache")));
+    }
+}
+
+#[async_trait::async_trait]
+impl BalanceFetching for CachingBalanceFetcher {
+    async fn get_balances(&self, queries: &[Query]) -> Vec<Result<U256>> {
+        let CacheResponse {
+            mut cached,
+            missing,
+            block: initial_block,
+        } = self.get_cached_balances(queries);
+
+        let missing_queries: Vec<Query> = missing.iter().map(|i| queries[*i]).collect();
+        let new_balances = self.inner.get_balances(&missing_queries).await;
+
+        let updates = missing
+            .iter()
+            .zip(new_balances.iter())
+            .filter_map(|(i, result)| Some((queries[*i], *result.as_ref().ok()?)))
+            .collect();
+        Self::update_balance_cache(&self.balance_cache, initial_block, updates);
+
+        cached.extend(missing.into_iter().zip(new_balances.into_iter()));
+        cached.sort_by_key(|(i, _)| *i);
+        cached.into_iter().map(|(_, balance)| balance).collect()
```

Comment:
Correct me if I'm wrong, but I think the cache grows in an unbounded way.

Reply:
It feels wasteful to have auto update on block and no expiry. Over time there will be lots of unused balance updates. If a balance hasn't been requested in X minutes it could drop from the cache.

For the size we don't need a limit but it would be good to have  metric for it.

Reply:
We should be able to use the `cached` crate for this (which has a similar interface to `HashMap` and shouldn't have an impact on the complexity here).